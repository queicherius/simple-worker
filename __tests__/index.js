const uuid = require('uuid')
const SimpleWorker = require('../src/index')

function filterLogData (data) {
  if (data && data.duration) {
    data.duration = 9001
  }

  if (data && data.data && data.data.errorStack) {
    data.data.errorStack = 'Somewhere over the rainbows'
  }

  return data
}

const makeTestQueue = (jobs) => {
  let logs = []

  const queue = new SimpleWorker({
    name: uuid(),
    redis: {
      host: '127.0.0.1',
      port: 6379
    },
    jobs: jobs,
    logger: {
      info: (message, data) => logs.push({type: 'info', message, data: filterLogData(data)}),
      warn: (message, data) => logs.push({type: 'warn', message, data: filterLogData(data)}),
      error: (message, data) => logs.push({type: 'error', message, data: filterLogData(data)})
    }
  })

  queue._logs = logs

  return queue
}

async function sleep (ms) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

describe('SimpleWorker', () => {
  test('has PRIORITIES', () => {
    expect(Object.keys(SimpleWorker.PRIORITIES)).toEqual(['HIGH', 'MEDIUM', 'LOW'])
  })

  test('can create a new queue instance', () => {
    const queue = makeTestQueue([])
    expect(queue.add).not.toBe(undefined)
    expect(queue._logs).toEqual([])
  })

  it('can queue and subsequently process a job', async () => {
    // Create a job to test if logging and passing in data is working
    let jobWasProcessed = false
    const jobFunction = async (job) => {
      job.info('Starting job function')
      await sleep(1000)
      jobWasProcessed = job.data.input
      job.warn('Finished job function')
    }

    // Create the queue with the job options
    const queue = makeTestQueue([{
      name: 'testing-processing',
      handler: jobFunction,
      options: {
        priority: SimpleWorker.PRIORITIES.HIGH,
        timeout: 10000
      }
    }])

    // Add the job to the queue
    queue.add('testing-processing', {input: '123'})

    // Check if the job is in the queue
    await sleep(250)
    expect((await queue.list()).map(job => job.data)).toMatchSnapshot()

    // Start processing and check if it is in the list still
    queue.process()
    await sleep(250)
    expect((await queue.list()).map(job => job.data)).toMatchSnapshot()
    expect(queue._logs).toMatchSnapshot()

    // Wait until the job should be done
    await sleep(1000)
    expect((await queue.list()).map(job => job.data)).toEqual([])
    expect(queue._logs).toMatchSnapshot()
    expect(jobWasProcessed).toEqual('123')
  })

  it('logs errors in the queue', () => {
    const queue = makeTestQueue([])
    const error = new Error('Oh no.')
    queue._queue.emit('error', error, 'error running queue')
    expect(queue._logs.map(filterLogData)).toMatchSnapshot()
  })
})
