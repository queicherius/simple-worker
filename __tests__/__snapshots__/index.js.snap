// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleWorker can flush the queue 1`] = `
Array [
  Object {
    "handler": "testing-processing",
    "input": "123",
  },
  Object {
    "handler": "testing-processing",
    "input": "456",
  },
  Object {
    "handler": "testing-processing",
    "input": "789",
  },
]
`;

exports[`SimpleWorker can queue and subsequently process a job 1`] = `
Array [
  Object {
    "handler": "testing-processing",
    "input": "123",
  },
]
`;

exports[`SimpleWorker can queue and subsequently process a job 2`] = `
Array [
  Object {
    "handler": "testing-processing",
    "input": "123",
  },
]
`;

exports[`SimpleWorker can queue and subsequently process a job 3`] = `
Array [
  Object {
    "data": Object {
      "data": Object {
        "handler": "testing-processing",
        "input": "123",
      },
      "name": "testing-processing",
    },
    "message": "Adding new job to the queue",
    "type": "info",
  },
  Object {
    "data": undefined,
    "message": "Starting job processing",
    "type": "info",
  },
  Object {
    "data": Object {
      "attempt": 0,
      "data": Object {
        "input": "123",
      },
      "jobId": "1",
      "name": "testing-processing",
    },
    "message": "Job processing started",
    "type": "info",
  },
  Object {
    "data": Object {
      "attempt": 0,
      "data": Object {
        "input": "123",
      },
      "jobId": "1",
      "messageData": undefined,
      "name": "testing-processing",
    },
    "message": "Starting job function",
    "type": "info",
  },
]
`;

exports[`SimpleWorker can queue and subsequently process a job 4`] = `
Array [
  Object {
    "data": Object {
      "data": Object {
        "handler": "testing-processing",
        "input": "123",
      },
      "name": "testing-processing",
    },
    "message": "Adding new job to the queue",
    "type": "info",
  },
  Object {
    "data": undefined,
    "message": "Starting job processing",
    "type": "info",
  },
  Object {
    "data": Object {
      "attempt": 0,
      "data": Object {
        "input": "123",
      },
      "jobId": "1",
      "name": "testing-processing",
    },
    "message": "Job processing started",
    "type": "info",
  },
  Object {
    "data": Object {
      "attempt": 0,
      "data": Object {
        "input": "123",
      },
      "jobId": "1",
      "messageData": undefined,
      "name": "testing-processing",
    },
    "message": "Starting job function",
    "type": "info",
  },
  Object {
    "data": Object {
      "attempt": 0,
      "data": Object {
        "input": "123",
      },
      "jobId": "1",
      "messageData": undefined,
      "name": "testing-processing",
    },
    "message": "Sleep finished",
    "type": "warn",
  },
  Object {
    "data": Object {
      "attempt": 0,
      "data": Object {
        "input": "123",
      },
      "jobId": "1",
      "messageData": undefined,
      "name": "testing-processing",
    },
    "message": "Finished job function",
    "type": "error",
  },
  Object {
    "data": Object {
      "attempt": 0,
      "data": Object {
        "input": "123",
      },
      "duration": 9001,
      "jobId": "1",
      "name": "testing-processing",
      "result": undefined,
    },
    "message": "Job processing completed",
    "type": "info",
  },
]
`;

exports[`SimpleWorker logs errors in the queue 1`] = `
Array [
  Object {
    "data": Object {
      "errorMessage": "Oh no.",
      "errorStack": "Somewhere over the rainbows",
    },
    "message": "Queue error",
    "type": "error",
  },
]
`;

exports[`SimpleWorker logs stalled jobs in the queue 1`] = `
Array [
  Object {
    "data": Object {
      "attempt": 1,
      "data": Object {
        "email": "pepe@me.me",
      },
      "jobId": 333,
      "name": "test-handler",
    },
    "message": "Job processing stalled",
    "type": "warn",
  },
]
`;
